docs: documentation updates to abci (#18625)
cosmos/cosmos-sdk: docs/build/abci++/00-introduction.md

# Introduction

> ## What is ABCI++?

Reviewer: ```suggestion
## What is ABCI?
```
any chance we can call abci++ just abci? it will get confusing in the future if abci 3.0 comes out. Is it abci+++ or abci 3.0.

Reviewer 2: +1

> ABCI++, Application Blockchain Interface is the interface between CometBFT and the application. Within the release of ABCI 2.0 for the 0.38 CometBFT release there were additional methods introduced.

Reviewer: ```suggestion
ABCI, Application Blockchain Interface is the interface between CometBFT and the application, more information about ABCI can be found [here](link). Within the release of ABCI 2.0 for the 0.38 CometBFT release there were additional methods introduced.
```

The 5 methods introduced during ABCI 2.0 are:

- `PrepareProposal`
- `ProcessProposal`
- `ExtendVote`
- `VerifyVoteExtension`
- `FinalizeBlock`

## The Flow

## PrepareProposal

Based on their voting power, CometBFT chooses a block proposer and calls `PrepareProposal` on the block proposer's application (Cosmos SDK). The selected block proposer is responsible for collecting outstanding transactions from the mempool, adhering to the application's specifications. The application can enforce custom transaction ordering and incorporate additional transactions, potentially generated from vote extensions in the previous block.

> To perform this manipulation on the application side, a custom handler is implemented. If no custom handler exists, all transactions are validated. Please see [this](https://github.com/fatal-fruit/abci-workshop) tutorial for more information on custom handlers.

Reviewer: will this tutorial get moved to tutorials? its a bit weird we link to a random repo here. Could be confusing for users

Author: yep, thats the plan for this week but needed to get something out asap for people to use. Its only for the interim. Following moving the tutorial across this link will point to the `sdk-tutorials` repo. I mentioned here that the repo will be moved across https://github.com/cosmos/cosmos-sdk-docs/pull/91 but should of made mention here too.

> To perform this manipulation on the application side, a custom handler is implemented. If no custom handler exists, all transactions are validated. Please see [this](https://github.com/fatal-fruit/abci-workshop) tutorial for more information on custom handlers.

Reviewer: ```suggestion
To perform this manipulation on the application side, a custom handler must implemented. By default, the Cosmos SDK provides a `PrepareProposalHandler`, which if used with a application specific mempool. A custom handler can be written by application developer, if a noop handler provided, all transactions are considered valid. Please see [this](https://github.com/fatal-fruit/abci-workshop) tutorial for more information on custom handlers.
```

Reviewer 3: > To perform this manipulation on the application side, a custom handler must be implemented. By default, the Cosmos SDK provides a `PrepareProposalHandler`, >>which if used with a application specific mempool<<. A custom handler can be written by the application developer, but if a noop handler is provided, all transactions are considered valid

I added some connectors that I think were missing, I still don't know what you meant with "which if used with a application specific mempool"

> Please note that vote extensions will only be available on the following height in which vote extensions are enabled. To see more details on vote extensions please see [here](https://docs.cosmos.network/main/build/abci++/01-vote-extensions).

Reviewer: ```suggestion
Please note that vote extensions will only be available on the following height in which vote extensions are enabled. More information about vote extensions can be found [here](https://docs.cosmos.network/main/build/abci++/01-vote-extensions).
```
